### PR TITLE
[SM-654] Add support for direct secret permissions at the repo layer

### PR DESCRIPTION
**Pull request by Thomas-Avery**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/SM-654

## 📔 Objective

The objective of this PR is to update the Secrets Manager repository layer to honor direct secret permissions.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: test/Api.IntegrationTest/SecretsManager/Controllers/ServiceAccountsControllerTests.cs [hunk 1]
```diff
@@ -73,51 +77,94 @@ public async Task ListByOrganization_SmAccessDenied_NotFound(bool useSecrets, bo
...
+                var serviceAccountResult = result.Data.FirstOrDefault(x => x.Id == item.Key);
+                if (serviceAccountResult != null)
+                {
+                    Assert.Equal(item.Value, serviceAccountResult.AccessToSecrets);
+                }
+            }
+        }
+        else
+        {
+            Assert.DoesNotContain(result.Data, x => x.AccessToSecrets != 0);
```
Comment (mzieniukbw) [inline test/Api.IntegrationTest/SecretsManager/Controllers/ServiceAccountsControllerTests.cs:166]: ⛏️ i thin; it's better to not use double negation, easier to read like below:
```suggestion
            Assert.Contains(result.Data, x => x.AccessToSecrets == 0);
```

**Review by mzieniukbw — COMMENTED** [review]
Looks good, got few minor comments though.

File: bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/SecretRepository.cs [hunk 2]
```diff
@@ -434,4 +420,40 @@ await dbContext.ServiceAccount
...
+                s.GroupAccessPolicies.Any(ap =>
+                    ap.Group.GroupUsers.Any(gu => gu.OrganizationUser.User.Id == userId && ap.Write)) ||
+                s.Projects.Any(p =>
+                    p.UserAccessPolicies.Any(ap => ap.OrganizationUser.User.Id == userId && ap.Write) ||
+                    p.GroupAccessPolicies.Any(ap =>
+                        ap.Group.GroupUsers.Any(gu => gu.OrganizationUser.User.Id == userId && ap.Write)))
+            )),
+            AccessClientType.ServiceAccount => secrets.Select(s => new SecretAccess(
+                s.Id,
+                s.ServiceAccountAccessPolicies.Any(ap => ap.ServiceAccountId == userId && ap.Read) ||
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/SecretRepository.cs]: Renamed to `accessClientId` https://github.com/bitwarden/server/pull/4156/commits/57bf4b190a4d3b1855d7dc994c8fd5d16f288da5

File: bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/SecretRepository.cs [hunk 3]
```diff
@@ -434,4 +420,40 @@ await dbContext.ServiceAccount
...
+            AccessClientType.User => secrets.Select(s => new SecretAccess(
+                s.Id,
+                s.UserAccessPolicies.Any(ap => ap.OrganizationUser.User.Id == userId && ap.Read) ||
+                s.GroupAccessPolicies.Any(ap =>
+                    ap.Group.GroupUsers.Any(gu => gu.OrganizationUser.User.Id == userId && ap.Read)) ||
+                s.Projects.Any(p =>
+                    p.UserAccessPolicies.Any(ap => ap.OrganizationUser.User.Id == userId && ap.Read) ||
+                    p.GroupAccessPolicies.Any(ap =>
+                        ap.Group.GroupUsers.Any(gu => gu.OrganizationUser.User.Id == userId && ap.Read))),
+                s.UserAccessPolicies.Any(ap => ap.OrganizationUser.User.Id == userId && ap.Write) ||
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/SecretRepository.cs]: I tried several ways, but it doesn't seem possible to do this in an EF projection query. 

We can pass in an expression to the Select `private Expression<Func<Secret, SecretAccess>> SecretToSecretAccess()` but that won't achieve the one function / lambda envisioned.


An expression within the expression doesn't work.

```
private Expression<Func<Secret, bool>> UserAccess(bool read, bool write)`
```

```
     AccessClientType.User => secrets.Select(s => new SecretAccess(
                s.Id,
                UserAccess(true, false),
                UserAccess(false, true),
            )),
```

Using a function would require us to do the filtering in C# and return a larger result set from the DB, which isn't desirable. https://stackoverflow.com/questions/6430182/using-function-in-select-clause-of-entity-framework-query

EF is just translating all this to SQL, so these lambdas aren't what is actually getting executed.

File: bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/ServiceAccountRepository.cs [hunk 4]
```diff
@@ -200,4 +199,48 @@ accessType switch
...
+    private static IQueryable<ServiceAccountSecretsAccess> BuildDirectSecretAccessQuery(
+        DatabaseContext dbContext,
+        IQueryable<ServiceAccount> serviceAccountQuery) =>
+        from sa in serviceAccountQuery
+        join ap in dbContext.ServiceAccountSecretAccessPolicy
+            on sa.Id equals ap.ServiceAccountId into grouping
+        from ap in grouping.DefaultIfEmpty()
+        where ap.GrantedSecret.DeletedDate == null
+        select new ServiceAccountSecretsAccess(sa,
+            ap.GrantedSecretId.HasValue
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/ServiceAccountRepository.cs]: Updated in https://github.com/bitwarden/server/pull/4156/commits/2c565d448cf2ca74e371a733cba3120d1c852c16

File: bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/ServiceAccountRepository.cs [hunk 5]
```diff
@@ -200,4 +199,48 @@ accessType switch
...
+            var secretId = directSecretAccessResult.SecretIds.FirstOrDefault();
+            if (secretId == Guid.Empty)
+            {
+                return false;
+            }
+
+            var serviceAccountProjectsSecretsAccessResults = projectSecretsAccessResults
+                .Where(x => x.ServiceAccount.Id == serviceAccountId)
+                .ToList();
+            return !serviceAccountProjectsSecretsAccessResults.Any(x => x.SecretIds.Contains(secretId));
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/ServiceAccountRepository.cs]: Updated in https://github.com/bitwarden/server/pull/4156/commits/2c565d448cf2ca74e371a733cba3120d1c852c16

File: test/Api.IntegrationTest/SecretsManager/Controllers/ServiceAccountsControllerTests.cs [hunk 6]
```diff
@@ -73,51 +77,94 @@ public async Task ListByOrganization_SmAccessDenied_NotFound(bool useSecrets, bo
...
-        var accessPolicies = serviceAccountIds.Take(2).Select(
-            id => new UserServiceAccountAccessPolicy
+        Assert.NotNull(result);
+        Assert.NotEmpty(result.Data);
+        Assert.Equal(serviceAccountIds.Count, result.Data.Count());
+
+        foreach (var item in expectedAccess)
+        {
+            var serviceAccountResult = result.Data.FirstOrDefault(x => x.Id == item.Key);
+            if (serviceAccountResult != null)
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply test/Api.IntegrationTest/SecretsManager/Controllers/ServiceAccountsControllerTests.cs]: Updated in https://github.com/bitwarden/server/pull/4156/commits/c9e5d972170e1507c708060cf085e3a252d497a6

File: test/Api.IntegrationTest/SecretsManager/Controllers/ServiceAccountsControllerTests.cs [hunk 7]
```diff
@@ -73,51 +77,94 @@ public async Task ListByOrganization_SmAccessDenied_NotFound(bool useSecrets, bo
...
-        Assert.Equal(2, result.Data.Count());
+        Assert.Equal(serviceAccountIds.Count, result.Data.Count());
+        Assert.DoesNotContain(result.Data, x => x.Id == serviceAccountWithoutAccess.Id);
+
+        if (includeAccessToSecrets)
+        {
+            foreach (var item in expectedAccess)
+            {
+                var serviceAccountResult = result.Data.FirstOrDefault(x => x.Id == item.Key);
+                if (serviceAccountResult != null)
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply test/Api.IntegrationTest/SecretsManager/Controllers/ServiceAccountsControllerTests.cs]: Updated in https://github.com/bitwarden/server/pull/4156/commits/c9e5d972170e1507c708060cf085e3a252d497a6

File: test/Api.IntegrationTest/SecretsManager/Controllers/ServiceAccountsControllerTests.cs (hunk 1 above)
Comment (Thomas-Avery) [reply test/Api.IntegrationTest/SecretsManager/Controllers/ServiceAccountsControllerTests.cs]: Updated in https://github.com/bitwarden/server/pull/4156/commits/c9e5d972170e1507c708060cf085e3a252d497a6

**Review by mzieniukbw — APPROVED** [review]
Nice! LGTM

**Review by coltonhurst — APPROVED** [review]

**Pull request merged**